imp: update front

Mesteery opened the pull request:
Résous https://github.com/readthedocs-fr/bin/issues/29 et https://github.com/readthedocs-fr/bin/issues/34.

J'ai amélioré le front :

- ajout des contrôles pour l'expiration et de l'utilisation maximum
- meilleure responsivité
- suppression de fichiers inutiles
- simplifications de plusieurs choses

![Contrôles](https://user-images.githubusercontent.com/48163546/103136497-39f63000-46c1-11eb-9d79-828dafb956cb.png)
**Sur téléphone :** https://user-images.githubusercontent.com/48163546/103136967-17661600-46c5-11eb-85e3-80cd5dd0ac8c.jpg

J'ai également modifier le controller pour qu'il passe le `DEFAULT_LANGUAGE`  dans la vue, pour que le langage par défaut soit sélectionner par défaut. J'ai aussi rajouter des "hints" sur les contrôles (submit, expiration, max usage, language) :
![Hints](https://user-images.githubusercontent.com/48163546/103136566-c99bde80-46c1-11eb-8540-a805dcacaae4.png)

Je propose aussi de mieux séparer le CSS de `newform.html` et de `highlight.html`.

J'ai du utiliser un peu de JS (`loc.js`) pour régler le problème de numérotation. L'idéal serait de directement mettre la valeur (`colWidth`) (qui est égale au "nombre de chiffres du nombre de lignes" - 0.5) depuis le back dans le CSS.

**Avant (y a pas les scrollbars dans le screen) :**
https://user-images.githubusercontent.com/48163546/103136580-e506e980-46c1-11eb-8d1a-4f3857801c20.png
**Après (line-height = 150%) :**
https://user-images.githubusercontent.com/48163546/103136597-0b2c8980-46c2-11eb-9b2d-c8008a9426d2.png
**Après (line-height = 200%) :** (du coup je propose de mettre la line-height a 200%, elle est actuellement a 125%, 150% dans cette PR)
https://user-images.githubusercontent.com/48163546/103136601-1089d400-46c2-11eb-84b2-da56ed81d0a4.png

Je voulais aussi faire https://github.com/readthedocs-fr/bin/issues/9, mais malheureusement la structure actuelle ne le permet pas, car il est impossible d'écouter les cliques sur un pseudo-élément (on est obligé d'écouter sur son parent). Ca impliquerai de tout changer (ne plus utiliser des ordered lists et donc changer la structure).

Quand j'ai lancé l'unit test j'ai eu une erreur (je l'ai reglé) :
```
======================================================================
ERROR: tearDownClass (tests.test_controller.TestController)
----------------------------------------------------------------------
TypeError: tearDownClass() missing 1 required positional argument: 'self'

----------------------------------------------------------------------
Ran 5 tests in 0.544s

FAILED (errors=1)
```

NoeMastrorillo commented on bin/views/newform.html:
```diff
@@ -1,25 +1,45 @@
+          <option value="604800">1 semaine</option>
+          <option value="86400">1 jour</option>
+          <option value="3600">1 heure</option>
+          <option value="60">1 minute</option>
```
Julien a fait une lib qui permet d'écrire les métriques plus facilement. De plus, faire que le snippet expire dans 1 mois ou 1 an c'est inutile car beaucoup trop long. Et on avait dit que `-1` représentait l'infini.
En attendant que Julien ajoute les jours, on peut faire comme ceci :
```suggestion
          <option selected value="-1">Jamais</option>
          <!-- option value="8760h">1 an</option -->
          <!-- option value="720h">1 mois</option -->
          <option value="168h">1 semaine</option>
          <option value="24h">1 jour</option>
          <option value="1h">1 heure</option>
          <option value="1m">1 minute</option>
```

NoeMastrorillo replied:
Il faudra de plus l'adapter côté back:
```python
lifetime = int(Time(forms.get('lifetime')) or lifetime)
```

NoeMastrorillo replied:
Et pour finir, ça serait bien de pouvoir choisir combien de minutes, heures et jours on veut que le snippet dure ; à présent on est limités à 1 minute, 1 heure, 1 jour (je ne pense pas que ce soit utile de monter au dessus d'1 semaine).

Mesteery replied:
Oui oui je sais c'est exprès, pour pouvoir en discuter, et j'ai fait exprès de mettre les secondes en dure car le back supporte que ça.
D'ailleurs j'avais mis -1 je comprends pas pourquoi c'est 0

NoeMastrorillo replied:
> Oui oui je sais c'est exprès, pour pouvoir en discuter, et j'ai fait exprès de mettre les secondes en dure car le back supporte que ça.

Rien ne t'empêche de modif un peu le back même si c'est une PR front. Je t'ai d'ailleurs fournis le code nécessaire à implémenter niveau back donc à ce niveau aucun problème 👌

Mesteery replied:
À voir l'avis des autres

Mesteery replied:
Mais je vois pas trop l'intérêt la en vrai que ce sois écrit en dure ou en format ça change rien. Tant que le front est assez explicite, c'est très.

Mesteery replied:
Et si jamais c'est écrit en format par l'utilisateur, peut-être que ça devrait être converti par le front ?

NoeMastrorillo replied:
Au moins on reste constant et c'est plus lisible. Et aussi, si jamais on fait en sorte que l'utilisateur puisse choisir exactement la durée du snippet, il pourra entrer des valeurs comme par exemple "5m19s".

NoeMastrorillo replied:
Et puis ça coute rien à implémenter, c'est pas comme si on parlait de refactor tout le code en entier.

Mesteery replied:
À voir l'avis des autres, mais pour moi ça devrait être implémenté niveau front et non niveau back.

Julien00859 replied:
Je suis d'accord avec Noé, on a une lib pour parser efficacement des mesures de temps, autant en tirer profit, ça permettra aussi aux utilisateurs en ligne de commande via curl de pouvoir définir la lifetime via le format supporté par metrics. By the way, la conversion en `int` n'est pas nécessite, la classe Time hérite déjà de `int` si je ne m'abuse.

NoeMastrorillo commented on bin/views/newform.html:
```diff
@@ -1,25 +1,45 @@
-        </form>
-    </body>
+      <div class="controls">
+        <input class="control" type="number" min="-1" title="Nombre d'utilisation maximum" placeholder="Nombre d'utilisation maximum" name="maxusage" />
```
```suggestion
        <input class="control" type="number" min="-1" title="Nombre d'utilisations maximum" placeholder="Nombre d'utilisations maximum" name="maxusage" />
```

NoeMastrorillo replied:
Il faudrait de plus pouvoir choisir "infini" (-1 n'est pas très explicite) ; j'avais pensé à une checkbox qui quand elle est cochée définit la valeur sur "infini" et l'input est invisible, et quand elle est décochée elle rend l'input visible pour pouvoir entrer un nombre.

Mesteery replied:
C'est par défaut à infini, donc s'il veut que ce soit infini soit il laisse vide soit il met -1. Mais il faudrait qu'il sache que c'est optionnel et que c'est par défaut à infini.

NoeMastrorillo replied:
Ouep mais du coup je trouve que de faire une checkbox c'est plus explicite.

Mesteery replied:
À mon avis, une checkbox n'est vraiment pas la solution. Au pire des cas, une `datalist` avec du JS derrière

Julien00859 commented on bin/assets/styles/monokai.css:
```diff
@@ -3,14 +3,14 @@ generated by Pygments <https://pygments.org/>
 span.linenos.special { color: #000000; background-color: #ffffc0; padding-left: 5px; padding-right: 5px; }
 body .hll { background-color: #49483e }
 body .c { color: #75715e } /* Comment */
-body .err { color: #960050; background-color: #1e0010 } /* Error */
```
Pourquoi ?

Mesteery replied:
Je sais pas je me suis dis qu'en rouge ca rendrait mieux niveau UI

Julien00859 commented on bin/assets/styles/style.css:
```diff
@@ -1,108 +1,139 @@
-    width: inherit;
+.controls button.valid svg {
+  fill: #ffffff;
 }
```
Aucune chance que je laisse passer 127 changements sur un fichier qui faisait ~100 lignes à la base si il n'y a aucune justification technique concernant ces changements. Un message de commit c'est pas pour faire jolie.

Julien00859 commented on bin/controller.py:
```diff
@@ -40,7 +40,8 @@ def post_new():
             code = part.file.read(config.MAXSIZE)
-            ext = parse_extension(Path(part.filename).suffix.lstrip('.')) or ext
+            ext = parse_extension(
+                Path(part.filename).suffix.lstrip('.')) or ext
```
C'est très mal venu de corriger le code des autres si c'est *uniquement* pour faire un changement de style. Je n'applique pas à la lettre PEP8, j'essaie d'éviter de dépasser la limite des 79 caractères mais je n'hésite pas à le faire si le style en deviendrait lourdingue, ici l'avoir sur deux lignes c'est lourdingue.

Mesteery replied:
C'est pas voulu, a chaque fois j'essaye de Ctrl + K + Ctrl + Shift + S pour pas que prettier corrige, mais j'ai du oublier. En tout cas, c'était pas le but.

Julien00859 commented on bin/views/highlight.html:
```diff
@@ -1,10 +1,11 @@
-    <body>
-{{!codehl}}
-    </body>
+<html lang="fr">
```
Les quotes ne sont pas nécessaires autours des attributs en html

Mesteery replied:
https://github.com/readthedocs-fr/bin/pull/53#discussion_r548910670

Julien00859 commented on bin/views/highlight.html:
```diff
@@ -1,10 +1,11 @@
+<html lang="fr">
+  <head>
+    % include('head.html')
+    <link rel="stylesheet" href="/assets/styles/monokai.css" />
```
Le document est écrit en HTML5, pas en XHTML, il n'y a pas besoin de fermer les balises orphelines. `<link>` est tout à fait valide.

Mesteery replied:
https://github.com/readthedocs-fr/bin/pull/53#discussion_r548910670

Julien00859 commented on bin/views/newform.html:
```diff
@@ -1,25 +1,45 @@
-                <option value="{{ext}}">{{lang.title()}}</option>
-                % end
-            </select>
+<html lang="fr">
```
Pareil qu'au dessus, les quotes ne sont pas nécessaires

Mesteery replied:
https://github.com/readthedocs-fr/bin/pull/53#discussion_r548910670

Julien00859 commented on bin/views/newform.html:
```diff
@@ -1,25 +1,45 @@
+          % end
+        </select>
+
+        <button class="control" title="Soumettre" type="submit">
```
Les quotes ne sont pas nécessaires. Mais IMO c'est bien d'utiliser des quotes pour des strings affichés à l'utilisateur et les classes, c'est surtout type= qui m'ennuie

Mesteery replied:
Il me semble que il était la depuis le début, et j'y est pas trop touche.

Julien00859 commented on bin/views/newform.html:
```diff
@@ -1,25 +1,45 @@
+              fill-rule="evenodd"
+              d="M2 6a2 2 0 012-2h4a1 1 0 010 2H4v10h10v-4a1 1 0 112 0v4a2 2 0 01-2 2H4a2 2 0 01-2-2V6z"
+              clip-rule="evenodd"
+            />
```
Pas besoin de changer le style pour rien, dans le jargon on appelle ça fait "un petit pipi", c'est très mal venu.

Mesteery replied:
https://github.com/readthedocs-fr/bin/pull/53#discussion_r548910670

Julien00859 commented on tests/test_controller.py:
```diff
@@ -120,6 +120,7 @@ def setUpClass(cls):
         bottle.template = MagicMock()
         bottle.template.side_effect = bottle_template
 
+    @classmethod
```
Il faut aussi changer le nom de l'argument, de `self` vers `cls`. Merci pour ce correctif !

Julien00859 commented on bin/assets/scripts/newform-button.js:
```diff
@@ -33,9 +25,11 @@ window.addEventListener("keydown", (event) => {
 
     // inserts tab at the position of the caret
-    code.value = content.substring(0, pos) + '\t' + content.substring(pos);
+    code.value = content.substring(0, pos) + "\t" + content.substring(pos);
```
Ça ne se fait pas de changer le code des autres pour rien

Mesteery replied:
https://github.com/readthedocs-fr/bin/pull/53#discussion_r548910670

Julien00859 reviewed (changes requested):
Il y a une profonde refonte du CSS mais aucune justification de pourquoi ces changements sont nécessaire, les messages de commits doivent être revu pour expliquer en long en large la nature des changements et pourquoi c'est mieux ainsi qu'avant.

Je suis aussi d'accord avec la proposition de Noé, je trouve qu'utiliser Time dans le backend serait un ajout judicieux, pourquoi limiter cette fonctionnalité au frontend ? ;)

Sinon globalement, c'est **très** mal venu (surtout pour une 1e contribution) de changer autant de code pour rien. Il y a de nombreux endroits où tu ne modifies les lignes *que* pour faire des petites adaptations de style, les quotes autours des valeurs des attributs html alors que ce n'est pas nécessaire, les balises orphelines qui d'un coup reçoivent un `/` inutile, du code qui est juste modifié pour plaire à tes goûts de développeur. Sache que tu n'es pas seul sur ce projet, ça fait deux mois qu'il existe, il y a pas mal de gens qui y ont déjà contribué, tu n'as tout simplement pas le droit de venir lâcher un petit pipi sur les contributions des autres.

Bonne idée pour avoir le `default_langage` dans le frontend :+1:

Mesteery committed:
imp: support durations with Time

We were forced to use seconds for the lifetime.
In the frontend, the language selected by default was the first in the list.

Now, durations (e.g., 1m, 1h, ...) are supported.
And the language selected by default is the default language.

Mesteery committed:
fix: test accept some tags as singles pairs

The test unit produced an error if the "option" item was not closed.

Now it accepts "option" and "optgroup" as singles pairs
if they are respectively direct children of datalist and select.

Co-authored-by: Julien Castiaux <julien.castiaux@gmail.com>

Mesteery committed:
imp: support lifetime and maxusage option

Closes: https://github.com/readthedocs-fr/bin/issues/34

The default selected language is now the default language defined in the backend.

Mesteery committed:
ref: refactor scripts

We use now the standard.js style guide.

Mesteery committed:
fix: lines number

The lines numbers could overflow if the lines were very numerous.
I fixed that with a bit of JavaScript

Mesteery committed:
ref: refactor styles

Some of the styles contained unnecessary code.

I added responsiveness, more padding and some micro changes.
I also removed the dot from the lines number and refactor the CSS.

Julien00859 commented on bin/assets/scripts/loc.js:
```diff
@@ -0,0 +1,2 @@
+const locWidth = document.getElementsByTagName('li').length.toString().length;
+document.documentElement.style.setProperty('--loc-width', `${locWidth - 0.5}em`);
```
Could you provide a screenshot of the related bug ?

Mesteery replied:
The bug with actual version :
![image](https://user-images.githubusercontent.com/48163546/103156767-9bdf9400-47ac-11eb-9ebf-3feea83c24b9.png)
https://rtdbin.fusetim.tk/yphawa.txt

The bug with my PR :
![image](https://user-images.githubusercontent.com/48163546/103156819-1c9e9000-47ad-11eb-90e7-0bd3489ef8a5.png)

Julien00859 reviewed (approved):
D'un point de vu du code, je pense qu'on est plutôt bon. Merci pour avoir refait ces messages de commits, ça m'a aidé à mieux comprendre les changements. J'aimerais avoir un avis supplémentaire pour le design.

Aussi, comme tu as fait pas mal de modification côté frontend html/css/js, autant pour le refectoring important du css qu'avoir appliqué les standards de style en js, tu deviendrais la personne ressource pour les changements côté frontend. Tu serais ok avec ça ?

NoeMastrorillo commented on bin/controller.py:
```diff
@@ -45,7 +46,7 @@ def post_new():
             ext = parse_extension(forms.get('lang')) or ext
             maxusage = int(forms.get('maxusage') or maxusage)
-            lifetime = int(forms.get('lifetime') or lifetime)
+            lifetime = Time(forms.get('lifetime') or lifetime)
```
Ici le lifetime par défaut est déjà un Time (voir config.py). De plus, pourquoi ne pas directement le convertir en int ici plutôt que dans la création du Snippet ?
```suggestion
            lifetime = int(Time(forms.get('lifetime')) or lifetime)
```

Mesteery replied:
`Time(forms.get('lifetime') or lifetime)` je suis obligé de faire comme ceci car si `lifetime` n'est pas défini, Time va renvoyer une erreur (voir https://gist.github.com/Julien00859/a63f6b830547059795689e549045853f#file-metrics-py-L134).
Pour le `int()` c'est un choix de Lazor que je ne saurais pas vraiment expliqué. Mais a mon avis, étant donné que ca hérite de int, et que on a juste besoin de le cast pour expire. (ici par exemple, https://github.com/readthedocs-fr/bin/pull/53/files#diff-d153f64554a9860b55fc20c353852e9116ae3d6d73b9d0387fda4ca07baec73cR23, y a pas besoin de cast, donc on cast vraiment que pour le expire)

Julien00859 replied:
Oui si on caste directement alors on perd le `__str__` custom des Time donc autant cast en int *uniquement* là où c'est nécessaire

NoeMastrorillo replied:
Ok, alors c'est bon pour moi 👌

Julien00859 replied:
Et c'est vrai qu'on pourrait `Time(form.get(...)) or DEFAULT_LIFETIME` mais pour ce que ça change....

Mesteery replied:
(non ca fait une erreur)

NoeMastrorillo commented on bin/models.py:
```diff
@@ -21,7 +21,7 @@ def create(cls, code, maxusage, lifetime):
         database.hset(ident, "views_left", maxusage)
         if lifetime > 0:
-            database.expire(ident, lifetime)
+            database.expire(ident, int(lifetime))
```
```suggestion
            database.expire(ident, lifetime)
```
